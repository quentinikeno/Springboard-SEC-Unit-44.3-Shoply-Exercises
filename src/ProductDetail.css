.ProductDetail .image img {
	width: fit-content;
}
