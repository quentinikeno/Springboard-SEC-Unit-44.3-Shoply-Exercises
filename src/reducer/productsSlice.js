import { createSlice } from "@reduxjs/toolkit";
import inventory from "../data.json";

const initialState = {
	inventory,
	cart: {},
};

export const productsSlice = createSlice({
	name: "products",
	initialState,
	reducers: {
		addToCart: (state, action) => {
			const { id, qty } = action.payload;
			const foundProduct = state.cart[id];
			if (foundProduct) {
				state["cart"][id]["quantity"] += qty;
			} else {
				state["cart"][id] = { quantity: qty };
			}
		},
		removeFromCart: (state, action) => {
			const { id, qty } = action.payload;
			const foundProduct = state.cart[id];
			if (foundProduct) {
				state["cart"][id]["quantity"] -= qty;
				if (state["cart"][id]["quantity"] < 1) delete state["cart"][id];
			} else {
				console.error("Item not found in cart.");
			}
		},
		clearCart: (state) => {
			state.cart = {};
		},
	},
});

export const { addToCart, removeFromCart, clearCart } = productsSlice.actions;

export default productsSlice.reducer;
