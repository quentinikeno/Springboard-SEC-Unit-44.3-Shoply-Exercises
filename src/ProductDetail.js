import { useSelector, useDispatch } from "react-redux";
import { addToCart, removeFromCart } from "./reducer/productsSlice";
import { useParams, Navigate, Link } from "react-router-dom";
import "./ProductDetail.css";

const ProductDetail = () => {
	const { id } = useParams();
	const product = useSelector(
		(state) => state.products.inventory.products[id]
	);
	const productInCart = useSelector((state) => state.products.cart[id]);
	const dispatch = useDispatch();

	if (!product) return <Navigate to="/" />;

	const { name, price, description, image_url } = product;

	const handleAddToCart = () => {
		dispatch(addToCart({ id, qty: 1 }));
	};

	const handleRemoveFromCart = () => {
		dispatch(removeFromCart({ id, qty: 1 }));
	};

	return (
		<div className="ProductDetail columns is-desktop my-3">
			<nav class="breadcrumb" aria-label="breadcrumbs">
				<ul>
					<li>
						<Link to="/">Products</Link>
					</li>
					<li class="is-active">
						<Link
							to="#"
							aria-current="page"
							className="is-capitalized"
						>
							{name}
						</Link>
					</li>
				</ul>
			</nav>
			<div className="column">
				<figure className="image is-4by3">
					<img src={image_url} className="mx-auto" />
				</figure>
			</div>
			<div className="column">
				<h1 className="title is-capitalized">{name}</h1>
				<p className="subtitle">${price}</p>
				<hr />
				<p className="box">
					<span className="has-text-weight-bold">Description: </span>
					<br /> {description}
				</p>
				<hr />
				<div>
					<button
						className="button is-primary is-rounded is-outlined is-fullwidth"
						onClick={handleAddToCart}
					>
						Add <i className="fa-solid fa-cart-plus ml-2"></i>
					</button>

					{productInCart && (
						<>
							<div className="box my-3">
								<span className="has-text-weight-bold">
									Number in Cart:{" "}
								</span>
								<br /> {productInCart.quantity}
							</div>

							<button
								className="button is-danger is-rounded is-outlined is-fullwidth"
								onClick={handleRemoveFromCart}
							>
								Remove{" "}
								<i className="fa-solid fa-cart-shopping ml-2"></i>
							</button>
						</>
					)}
				</div>
			</div>
		</div>
	);
};

export default ProductDetail;
