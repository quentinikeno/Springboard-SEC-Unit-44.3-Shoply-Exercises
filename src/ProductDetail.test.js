import { render } from "@testing-library/react";
import { store } from "./reducer/store";
import { Provider } from "react-redux";
import ProductDetail from "./ProductDetail";
import { MemoryRouter, Routes, Route } from "react-router-dom";

it("renders without crashing", () => {
	render(
		<Provider store={store}>
			<MemoryRouter
				initialEntries={[
					"/products/2266b2d6-3d71-4356-8065-392f1fceab07",
				]}
			>
				<Routes>
					<Route
						path="/products/:id"
						element={<ProductDetail />}
					></Route>
				</Routes>
			</MemoryRouter>
		</Provider>
	);
});

it("renders matches snapshot", () => {
	const { asFragment } = render(
		<Provider store={store}>
			<MemoryRouter
				initialEntries={[
					"/products/2266b2d6-3d71-4356-8065-392f1fceab07",
				]}
			>
				<Routes>
					<Route
						path="/products/:id"
						element={<ProductDetail />}
					></Route>
				</Routes>
			</MemoryRouter>
		</Provider>
	);
	expect(asFragment()).toMatchSnapshot();
});
