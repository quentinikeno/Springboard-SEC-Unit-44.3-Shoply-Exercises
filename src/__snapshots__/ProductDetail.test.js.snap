// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ProductDetail columns is-desktop my-3"
  >
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Products
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            class="is-capitalized"
            href="/products/2266b2d6-3d71-4356-8065-392f1fceab07"
          >
            phone
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="column"
    >
      <figure
        class="image is-4by3"
      >
        <img
          alt="phone"
          class="mx-auto"
          src="https://images-na.ssl-images-amazon.com/images/I/717DO5Q-aCL._SY879_.jpg"
        />
      </figure>
    </div>
    <div
      class="column"
    >
      <h1
        class="title is-capitalized"
      >
        phone
      </h1>
      <p
        class="subtitle"
      >
        $399.99
      </p>
      <hr />
      <p
        class="box"
      >
        <span
          class="has-text-weight-bold"
        >
          Description: 
        </span>
        <br />
         Because there are absolutely no downsides that come with easier access to social media.
      </p>
      <hr />
      <div>
        <button
          class="button is-primary is-rounded is-outlined is-fullwidth"
        >
          Add 
          <i
            class="fa-solid fa-cart-plus ml-2"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
